Enhance vmem allocator and slab cache

BrainSlayer opened the pull request:
this patch introduces several new features.
1. implementation of a kvmalloc like interface (kvmem_alloc/kvmem_zalloc)

2. fixes vmem_alloc implementation which can lead to slow allocations
since the first attempt with kmalloc does not make use of the noretry
flag but tells the linux kernel to retry several times before it fails.

3. implements KMC_KVMEM flag for kmem_cache_create which makes use of
a kvmalloc allocator implementation which may increase performance if
the allocator is able to allocate physical memory. otherwise it will
still make use of virtual memory.

4. make use of __GFP_HIGHMEM flag in vmem_alloc , which is required for
some 32 bit systems to make use of full available memory.

5. implement KM_ONCE flag which is behaving almost like KM_SLEEP but
does not hang in a deadloop if no memory is available. it should be used
if you want a GFP_KERNEL like memory allocation, but not that lazy like
KM_NOSLEEP which makes use of GFP_ATOMIC which is mainly made to be used
in irqcontexts. at some locations in zfs like compression algorithms
for instance GFP_ATOMIC/KM_NOSLEEP should not  be used since it might
always fail if blocks are greater than 64-128kb (workspaces). but we also
want no deadlock situation if memory is under pressure. so we need a way
that it behaves like straight GFP_KERNEL. this flag does allow this behaviour.

6. change vmem_alloc to kvmem_alloc on some usefull locations where
memory is initialized.

7. make use of KMC_KVMEM for spl_zlib_workspace_cache, ddt_cache,
zio_buf and zio_data_buf

see also: https://lwn.net/Articles/711653/

Signed-off-by: Sebastian Gottschall <s.gottschall@dd-wrt.com>

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

sempervictus commented:
First pass over that actually gave me a mild headache, one of those cases where reading the diff is pretty worthless as it lacks context.
In terms of the retry behavior, has there been any testing to see how it behaves under fragmentation pressure where it has to keep falling back to vmem_alloc? Wondering if there are any "zfs-ish" deadlocks or hangs hiding in the code which this might expose.
I've not built a master branch for some time, but this seems worth it. Hopefully this weekend...
Thank you for the PR.

BrainSlayer commented:
i have done stress tests with memory constrained systems (512 mb ram, 
slow cpu etc.)
what i have found is that the current kmem_alloc implementation can 
cause deadlocks. but this is unrelated to my patch.
i can reproduce this deadlock also without my patch. it basicly is 
running into a problem if there is no memory available anymore. no 
physical and no virtual.
so kmem_alloc will loop forever. on the other hand. if a internal 
function demands memory and its not available the FS is not function 
anymore anyway.
this deadlock is also not critical since system remains responsive, it 
may applications hang which are using the fs.

my patch does not change much here. the kmem_alloc with KM_VMEM hasnt 
been changed, except that i simplified the code a little bit.
if vmem is requested, it will request  physical memory first but noretry 
flag instead of standard kmalloc. so allocation will be quicker if 
kmalloc fails
since otherwise kmalloc will cause a significant overhead if not enough 
physical memory is available. so we fall qicker back to vmalloc which 
was requested anyway
so code locations which do allocations very often, are getting a benefit 
here since allocation is faster

in addition the new mode KM_KVMEM has been implemented which does ignore 
the kmem_alloc_max boundary but always tries to allocate physical memory 
and falls back
to vmalloc if failed. this has also another performance benefit. virtual 
memory is allocated on first use and has a significant overhead due 
paging. so if possible physical memory should be always
prefered. i converted some static initialized memory allocations from 
VMEM to KVMEM.
KVMEM also does not run into a deadlocking loop since it may fail under 
memory pressure. it should be used with error handling.
in my point of view, if used it forces developers to think about error 
handling instead of using a barrier allocator which may wait forever if 
memory is running out.
so far i only converted uncritical static initializers at driver init to 
KVMEM.

initially i wrote this enhancement on top of the zstd compression 
support since zstd may allocate alot of memory for compression workspace 
and i found out that kmalloc is performing much better than vmalloc
even at bigger sizes and while doing research i found out that kvmalloc 
which was introduced in linux 4.12 is getting about heavily used on alot 
of kernel drivers as replacement for vmalloc.
kvmalloc is faster than vmalloc in case physical memory is available. in 
case physical memory isnt available, its fast as vmalloc before which no 
significant overhead.
read https://lwn.net/Articles/711653/ if you want to know more about the 
intention and behaviour of this function. its basicly the better way 
instead of doing
mem = kmalloc()
if (!mem)
     mem = vmalloc()

since kmalloc is a blocking call and might wait for memory getting 
available for a unspecified retry time. i converted also kmem_cache to 
make use of this new behaviour which may bringt also performance 
benefits to the spl's own slab cache since these allocations are used 
very frequent over running time. so its all about allocation latency and 
memory access performance at all


Am 08.02.2019 um 03:22 schrieb RageLtMan:
>
> First pass over that actually gave me a mild headache, one of those 
> cases where reading the diff is pretty worthless as it lacks context.
> In terms of the retry behavior, has there been any testing to see how 
> it behaves under fragmentation pressure where it has to keep falling 
> back to vmem_alloc? Wondering if there are any "zfs-ish" deadlocks or 
> hangs hiding in the code which this might expose.
> I've not built a master branch for some time, but this seems worth it. 
> Hopefully this weekend...
> Thank you for the PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384#issuecomment-461668515>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHVmk8e_09XGvUFo_5H3ec1T5ecnC6Rpks5vLN9zgaJpZM4ameRr>.
>

c0d3z3r0 commented:
Successfully tested on Debian testing with upstream kernel (4.20.9 - 4.20.17); I have this merged for some weeks and did not have any problems.

BrainSlayer commented:
i will rework this for current git revision today since there is a small easy to resolve conflict. but thanks for testing

c0d3z3r0 commented:
@BrainSlayer https://github.com/c0d3z3r0/zfs/tree/rebase/pull/8384_vmem

BrainSlayer commented:
i did this already in my zstd tree which has the same patch

c0d3z3r0 commented:
You should do a rebase on master instead of merging it because you don‘t want the merge commits to get pulled

> Am 24.03.2019 um 13:36 schrieb Sebastian Gottschall <notifications@github.com>:
> 
> i did this already in my zstd tree which has the same patch
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

BrainSlayer commented:
never did this. git is way to complex and cryptic to be my favorite 
tool. the merge works too anyway

Am 24.03.2019 um 13:54 schrieb Michael Niewöhner:
> You should do a rebase on master instead of merging it because you 
> don‘t want the merge commits to get pulled
>
> > Am 24.03.2019 um 13:36 schrieb Sebastian Gottschall 
> <notifications@github.com>:
> >
> > i did this already in my zstd tree which has the same patch
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384#issuecomment-475956077>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHVmk8oAmHNVzfgMzRzS13m72lwjuuGCks5vZ3V9gaJpZM4ameRr>.
>

c0d3z3r0 commented:
#8382

c0d3z3r0 commented:
This improves performance a bit and at least for me did not introduce any problems so far on my machine. Anyone willing to review this? @kpande @behlendorf

behlendorf commented:
Sorry about the sluggish review.  The thrust of these changes look good, but I'd like to hold of fully reviewing and integrating them until after 0.8 is tagged.  After 0.8 is branched we can get these changes some soak time in master, and backport them if needed to 0.8.x.

BrainSlayer commented:
consider that this patch is also part of the zstd patch. its a spinoff 
of this project since zstd may allocate much bigger blocks for 
compression workspace which has alot of benefits
of this allocator modifcation.
but please think about the missing GFP_HIGHMEM flag in 
spl_kmem_alloc_impl. from my point of view this is a bug
and may cause uneccessary mem alloc fails on 32 bit systems. (the 
highmem boundary for some mips32 systems may be just 128 mb and is 
variable on x86 depending on the kernel config)
so i think that this small fix should be merged right now into the 
mainline and maybe also backported to 0.7


Am 11.04.2019 um 19:28 schrieb Brian Behlendorf:
>
> Sorry about the sluggish review. The thrust of these changes look 
> good, but I'd like to hold of fully reviewing and integrating them 
> until after 0.8 is tagged. After 0.8 is branched we can get these 
> changes some soak time in master, and backport them if needed to 0.8.x.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384#issuecomment-482216268>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHVmkw8Qb3JZFc1pk83Fiz6hChG4nPR0ks5vf3CugaJpZM4ameRr>.
>

behlendorf commented:
@BrainSlayer can you open a PR with just the minimal `GFP_HIGHMEM` fix so it can be reviewed independently and considered for inclusion.

BrainSlayer commented:
grrr. another fork of a fork of a fork :-)

Am 11.04.2019 um 19:53 schrieb Brian Behlendorf:
>
> @BrainSlayer <https://github.com/BrainSlayer> can you open a PR with 
> just the minimal |GFP_HIGHMEM| fix so it can be reviewed independently 
> and considered for inclusion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384#issuecomment-482225415>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHVmkxA1bXly9PJhehQ4uBsLScyMM6Vtks5vf3aZgaJpZM4ameRr>.
>

c0d3z3r0 commented:
> grrr. another fork of a fork of a fork :-) Am 11.04.2019 um 19:53 schrieb Brian Behlendorf:

You don't need to fork. This is what branches are made for ;)

BrainSlayer commented:
then its just a branch for a 8 characters patch which takes a long time 
to setup due the limited capablities of github. consider that my master 
tree is the zstd zfs tree and not the
zfs master. so creating a new branch from a different base is a little 
bit quirky

Am 11.04.2019 um 20:18 schrieb Michael Niewöhner:
>
>     grrr. another fork of a fork of a fork :-) Am 11.04.2019 um 19:53
>     schrieb Brian Behlendorf:
>
> You don't need to fork. This is what branches are made for ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384#issuecomment-482239416>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHVmk7RQHMix-XrZdN4KGBxW3gK_zjaNks5vf3xxgaJpZM4ameRr>.
>

c0d3z3r0 commented:
> then its just a branch for a 8 characters patch which takes a long time to setup due the limited capablities of github. consider that my master tree is the zstd zfs tree and not the zfs master. so creating a new branch from a different base is a little bit quirky

Uhm. What? You need to branch off of `master`, then apply your patch, commit, push, pr.
Leave this PR as-is. After the second PR got merged, do a `git rebase origin/master` in this PR's branch.

Example:

~~~sh
git fetch  # update

git branch gfp_highmen origin/master
git checkout gfp_highmen
vim ....   # do your changes
git commit <whatever>
git push -u origin

# wait four gfp_highmem to get merged

git fetch  # update again
git checkout vmem_alloc
git rebase origin/master
git push -f origin vmem_alloc
~~~

sempervictus commented:
So it boots, upgrades 0.7 test pools in VM, but causes hard lockups in the VM after a while in zloop (no stack trace even to ttyS0), and might also be leaking some kernel addresses via:
```
[Mon Jul  8 07:22:24 2019]  [<ffffffff819c1e0a>] ? pointer+0x47a/0x650
[Mon Jul  8 07:22:24 2019]  [<ffffffff819c222c>] ? vsnprintf+0x24c/0x5d0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8138027e>] ? __dprintf+0xfe/0x160
[Mon Jul  8 07:22:24 2019]  [<ffffffff81252c7b>] ? spl_kmem_zalloc+0x9b/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff8131d571>] ? metaslab_sync+0x991/0xba0
[Mon Jul  8 07:22:24 2019]  [<ffffffff81252c00>] ? spl_kmem_zalloc+0x20/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff8134d05a>] ? vdev_sync+0x6a/0x1d0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8132fa45>] ? spa_sync+0x5f5/0xf20
[Mon Jul  8 07:22:24 2019]  [<ffffffff8134402e>] ? spa_txg_history_init_io+0xfe/0x110
[Mon Jul  8 07:22:24 2019]  [<ffffffff81347885>] ? txg_sync_thread+0x2b5/0x480
[Mon Jul  8 07:22:24 2019]  [<ffffffff813475d0>] ? txg_thread_exit.isra.0+0xa0/0xa0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8125a67d>] ? thread_generic_wrapper+0x6d/0x80
[Mon Jul  8 07:22:24 2019]  [<ffffffff8125a610>] ? __thread_exit+0x10/0x10
[Mon Jul  8 07:22:24 2019]  [<ffffffff81088fe3>] ? kthread+0x113/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff81088ed0>] ? kthread_park+0x80/0x80
[Mon Jul  8 07:22:24 2019]  [<ffffffff8100136c>] ? ret_from_fork+0x2c/0x40
[Mon Jul  8 07:22:29 2019] audit: type=1131 audit(1562570550.153:49): pid=1 ui
```
(automated catch in a grsec kernel - all credit for finding that goes to spender/paxteam).

BrainSlayer commented:
the stack trace you posted shows that its not caused by this patch. 
spl_kmem_zalloc is a unchanged function path. this patch here changes 
only vmem allocations.
consider also that this patch is tested in productive environment for 
many months together with the zstd patch even with hard load. never seen 
any lockups.
but of course lockups may always occur if the system has no memory left 
for any actions. and i mean really "no memory"

Am 09.07.2019 um 03:19 schrieb RageLtMan:
>
> So it boots, upgrades 0.7 test pools in VM, but causes hard lockups in 
> the VM after a while in zloop (no stack trace even to ttyS0), and 
> might also be leaking some kernel addresses via:
>
> |[Mon Jul 8 07:22:24 2019] [<ffffffff819c1e0a>] ? pointer+0x47a/0x650 
> [Mon Jul 8 07:22:24 2019] [<ffffffff819c222c>] ? vsnprintf+0x24c/0x5d0 
> [Mon Jul 8 07:22:24 2019] [<ffffffff8138027e>] ? __dprintf+0xfe/0x160 
> [Mon Jul 8 07:22:24 2019] [<ffffffff81252c7b>] ? 
> spl_kmem_zalloc+0x9b/0x130 [Mon Jul 8 07:22:24 2019] 
> [<ffffffff8131d571>] ? metaslab_sync+0x991/0xba0 [Mon Jul 8 07:22:24 
> 2019] [<ffffffff81252c00>] ? spl_kmem_zalloc+0x20/0x130 [Mon Jul 8 
> 07:22:24 2019] [<ffffffff8134d05a>] ? vdev_sync+0x6a/0x1d0 [Mon Jul 8 
> 07:22:24 2019] [<ffffffff8132fa45>] ? spa_sync+0x5f5/0xf20 [Mon Jul 8 
> 07:22:24 2019] [<ffffffff8134402e>] ? 
> spa_txg_history_init_io+0xfe/0x110 [Mon Jul 8 07:22:24 2019] 
> [<ffffffff81347885>] ? txg_sync_thread+0x2b5/0x480 [Mon Jul 8 07:22:24 
> 2019] [<ffffffff813475d0>] ? txg_thread_exit.isra.0+0xa0/0xa0 [Mon Jul 
> 8 07:22:24 2019] [<ffffffff8125a67d>] ? 
> thread_generic_wrapper+0x6d/0x80 [Mon Jul 8 07:22:24 2019] 
> [<ffffffff8125a610>] ? __thread_exit+0x10/0x10 [Mon Jul 8 07:22:24 
> 2019] [<ffffffff81088fe3>] ? kthread+0x113/0x130 [Mon Jul 8 07:22:24 
> 2019] [<ffffffff81088ed0>] ? kthread_park+0x80/0x80 [Mon Jul 8 
> 07:22:24 2019] [<ffffffff8100136c>] ? ret_from_fork+0x2c/0x40 [Mon Jul 
> 8 07:22:29 2019] audit: type=1131 audit(1562570550.153:49): pid=1 ui |
>
> (automated catch in a grsec kernel - all credit for finding that goes 
> to spender/paxteam).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNEZSRTTVQAHKJ6OASO3P6PRSPA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZOZINQ#issuecomment-509449270>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE3FKHLORLCOAEACEILP6PRSPANCNFSM4GUZ4RVQ>.
>

c0d3z3r0 commented:
@BrainSlayer could you please create another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf, so we can (try to) get stuff merged one-by-one

BrainSlayer commented:
Am 09.07.2019 um 22:06 schrieb Michael Niewöhner:
>
> @BrainSlayer <https://github.com/BrainSlayer> could you please create 
> another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf 
> <https://github.com/behlendorf>, so we can (try to) get stuff merged 
> one-by-one
>
pff. thats a one liner. need to setup a new git tree first for that patch
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE4HGZETOWXFONGNOJ3P6TVTLA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZRL4KI#issuecomment-509787689>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNEYDN4X6V5F5Z3ATTPLP6TVTLANCNFSM4GUZ4RVQ>.
>

c0d3z3r0 commented:
No... just a branch ;) it only takes 1 minute of your time :)

> Am 12.07.2019 um 09:51 schrieb Sebastian Gottschall <notifications@github.com>:
> 
> 
> Am 09.07.2019 um 22:06 schrieb Michael Niewöhner:
> >
> > @BrainSlayer <https://github.com/BrainSlayer> could you please create 
> > another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf 
> > <https://github.com/behlendorf>, so we can (try to) get stuff merged 
> > one-by-one
> >
> pff. thats a one liner. need to setup a new git tree first for that patch
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub 
> > <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE4HGZETOWXFONGNOJ3P6TVTLA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZRL4KI#issuecomment-509787689>, 
> > or mute the thread 
> > <https://github.com/notifications/unsubscribe-auth/AB2WNEYDN4X6V5F5Z3ATTPLP6TVTLANCNFSM4GUZ4RVQ>.
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

c0d3z3r0 commented:
@BrainSlayer are you sure the highmem flag is really still needed? just found this one... https://github.com/torvalds/linux/commit/19809c2da28aee5860ad9a2eff760730a0710df0

BrainSlayer commented:
i assume no. for newer kernel revisions. but that doesnt matter we 
should ignore old kernels. whats the minimum kernel requirement?

Am 12.07.2019 um 19:17 schrieb Michael Niewöhner:
>
> @BrainSlayer <https://github.com/BrainSlayer> are you sure the highmem 
> flag is really still needed? just found this one... 
> torvalds/linux@19809c2 
> <https://github.com/torvalds/linux/commit/19809c2da28aee5860ad9a2eff760730a0710df0>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE5ZKJO3P6GH7PCUTGDP7C4A3A5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ2LGMQ#issuecomment-510964530>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE2GLPHWOLKKTRHBDE3P7C4A3ANCNFSM4GUZ4RVQ>.
>

c0d3z3r0 commented:
Minimum is 4.12-rc1. You mesnt „should **not** ignore old kernels“, didn‘t you?

BrainSlayer commented:
4.12-rc1 is clearly not the minimum requirement for zfs. i run it on 
5.2, 4.9, 4.4, 3.18 and even 3.10

Am 13.07.2019 um 05:54 schrieb Michael Niewöhner:
>
> Minimum is 4.12-rc1. You mesnt „should *not* ignore old kernels“, 
> didn‘t you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNEYM4PRCZPJO54I7SKDP7FGXRA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3IXGA#issuecomment-511085464>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE4AWH7AERMKHO7H5C3P7FGXRANCNFSM4GUZ4RVQ>.
>

c0d3z3r0 commented:
Yep, then we should include it.. thanks ;)

> Am 13.07.2019 um 06:01 schrieb Sebastian Gottschall <notifications@github.com>:
> 
> 4.12-rc1 is clearly not the minimum requirement for zfs. i run it on 
> 5.2, 4.9, 4.4, 3.18 and even 3.10
> 
> Am 13.07.2019 um 05:54 schrieb Michael Niewöhner:
> >
> > Minimum is 4.12-rc1. You mesnt „should *not* ignore old kernels“, 
> > didn‘t you?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub 
> > <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNEYM4PRCZPJO54I7SKDP7FGXRA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3IXGA#issuecomment-511085464>, 
> > or mute the thread 
> > <https://github.com/notifications/unsubscribe-auth/AB2WNE4AWH7AERMKHO7H5C3P7FGXRANCNFSM4GUZ4RVQ>.
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

c0d3z3r0 commented on include/spl/sys/kmem.h:
```diff
@@ -41,10 +47,12 @@ extern void strfree(char *str);
 #define	KM_SLEEP	0x0000	/* can block for memory; success guaranteed */
 #define	KM_NOSLEEP	0x0001	/* cannot block for memory; may fail */
 #define	KM_PUSHPAGE	0x0004	/* can block for memory; may use reserve */
+#define	KM_ONCE		0x0008	/* like KM_SLEEP but do not loop allocation */
```
Where does that get used? I couldn't find any usage here or in your zstd branch

c0d3z3r0 replied:
oops... should have been a comment, not a review....

BrainSlayer commented:
spl-kmem.c

its not used yet. but may be used. allocating with KM_SLEEP is common in 
zfs, but it has no error handling. under low memory conditions
KM_SLEEP may end in a deadlock / endless loop. situation. using KM_ONCE 
does allow to implement error handling at various locations in zfs if 
required.
so its a usefull api enhancement for future development

Am 13.07.2019 um 07:35 schrieb Michael Niewöhner:
>
> *@c0d3z3r0* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In include/spl/sys/kmem.h 
> <https://github.com/zfsonlinux/zfs/pull/8384#discussion_r303192851>:
>
> > @@ -41,10 +47,12 @@ extern void strfree(char *str);
>   #define	KM_SLEEP	0x0000	/* can block for memory; success guaranteed */
>   #define	KM_NOSLEEP	0x0001	/* cannot block for memory; may fail */
>   #define	KM_PUSHPAGE	0x0004	/* can block for memory; may use reserve */
> +#define	KM_ONCE		0x0008	/* like KM_SLEEP but do not loop allocation */
>
> Where does that get used? I couldn't find any usage here or in your 
> zstd branch
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE2SGOBGDALCY2RVTHTP7FSTXA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6LJWHY#pullrequestreview-261528351>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE5YSU6LQRUZMRICTMLP7FSTXANCNFSM4GUZ4RVQ>.
>

BrainSlayer commented:
dont bother. you're likely living in the same country as me and it was 
early morning

Am 13.07.2019 um 07:36 schrieb Michael Niewöhner:
>
> *@c0d3z3r0* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In include/spl/sys/kmem.h 
> <https://github.com/zfsonlinux/zfs/pull/8384#discussion_r303192863>:
>
> > @@ -41,10 +47,12 @@ extern void strfree(char *str);
>   #define	KM_SLEEP	0x0000	/* can block for memory; success guaranteed */
>   #define	KM_NOSLEEP	0x0001	/* cannot block for memory; may fail */
>   #define	KM_PUSHPAGE	0x0004	/* can block for memory; may use reserve */
> +#define	KM_ONCE		0x0008	/* like KM_SLEEP but do not loop allocation */
>
> oops... should have been a comment, not a review....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE6D5HODBLDMUHUQ5G3P7FSVZA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6LJWLA#discussion_r303192863>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE2T3CO674Q2K34PFN3P7FSVZANCNFSM4GUZ4RVQ>.
>

c0d3z3r0 commented:
> spl-kmem.c its not used yet. but may be used. allocating with KM_SLEEP is common in zfs, but it has no error handling. under low memory conditions KM_SLEEP may end in a deadlock / endless loop. situation. using KM_ONCE does allow to implement error handling at various locations in zfs if required. so its a usefull api enhancement for future development

Well then, I assume, it should not a) be part of this PR and b) do not get merged until there is any useful implementation. @behlendorf ?

> dont bother. you're likely living in the same country as me and it was early morning

Indeed :D

>>@BrainSlayer could you please create another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf, so we can (try to) get stuff merged one-by-one
>
>pff. thats a one liner. need to setup a new git tree first for that patch

Done now in #9031

BrainSlayer commented:
thanks for the PR :-) (it wont get merged anyway like all other pull 
requests)

Am 13.07.2019 um 07:53 schrieb Michael Niewöhner:
>
>     spl-kmem.c its not used yet. but may be used. allocating with
>     KM_SLEEP is common in zfs, but it has no error handling. under low
>     memory conditions KM_SLEEP may end in a deadlock / endless loop.
>     situation. using KM_ONCE does allow to implement error handling at
>     various locations in zfs if required. so its a usefull api
>     enhancement for future development
>
> Well then, I assume, it should not a) be part of this PR and b) do not 
> get merged until there is any useful implementation. @behlendorf 
> <https://github.com/behlendorf> ?
>
>     dont bother. you're likely living in the same country as me and it
>     was early morning
>
> Indeed :D
>
>         @BrainSlayer <https://github.com/BrainSlayer> could you please
>         create another PR for the GFP_HIGHMEM stuff as suggested by
>         @behlendorf <https://github.com/behlendorf>, so we can (try
>         to) get stuff merged one-by-one
>
>     pff. thats a one liner. need to setup a new git tree first for
>     that patch
>
> Done now in #9031 <https://github.com/zfsonlinux/zfs/pull/9031>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zfsonlinux/zfs/pull/8384?email_source=notifications&email_token=AB2WNE66NYTAIZBIH2H2BV3P7FUWLA5CNFSM4GUZ4RV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3KJ4Y#issuecomment-511091955>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2WNE3CSQ3IZYW4EMJI5TLP7FUWLANCNFSM4GUZ4RVQ>.
>

Ornias1993 commented:
@BrainSlayer I guess this PR requires some edits/updates?
It seems some features are split off into different PR's and It seems at least your intro text doesn't reflect that... Or am I missing something?

c0d3z3r0 commented:
> @BrainSlayer I guess this PR requires some edits/updates?
> It seems some features are split off into different PR's and It seems at least your intro text doesn't reflect that... Or am I missing something?

this is superseeded by my fork; we decided to not integrate KM_ONCE now but when it's needed later; the rest is integrated into my PR

Ornias1993 commented:
@c0d3z3r0 I guess this can be closed then? Or replaced by an issue/todo for later...

c0d3z3r0 commented:
> @c0d3z3r0 I guess this can be closed then? Or replaced by an issue/todo for later...

well, the code is not lost when closed; when it's needed later it's easy to restore it. I'd say this can be closed for now. @BrainSlayer what do you say?

Ornias1993 commented:
@c0d3z3r0 Thats fine indeed, did so before... Closing archive-only PR's works out fine :)

BrainSlayer commented:
i still like the test system of zfs for improvements. this gets lost if we close the PR. so i can keep up to date my code and still see if there are new problems

Ornias1993 commented:
@BrainSlayer atleast update your PR explaination accordingly then ;)

BrainSlayer commented:
@Ornias1993 ah now i get it. yes this PR here is dead and unmaintained. i focus on the zstd patch and @c0d3z3r0 took over this part and has his own PR based on my work. i did not check which PR this here was about :-)

c0d3z3r0 commented:
@BrainSlayer then please click "Close"

Ornias1993 commented:
@BrainSlayer No problem, tried to be a little more diplomatic about it :P
And what @c0d3z3r0  says...
